File: PokerProtocol/src/protocol/observer/CommandObserver.java
package protocol.observer;

import utility.EventObserver;

public class CommandObserver<T extends ICommandListener> extends EventObserver<T>
{
    public void receiveSomething(String line)
    {
        if (line == null)
        {
            return;
        }
        baseReceive(line);
        commandReceived(line);
    }
    
    protected void commandReceived(String line)
    {
        
    }
    
    protected void baseReceive(String line)
    {
        for (final T listener : getSubscribers())
        {
            listener.commandReceived(line);
        }
    }
}

File: C#Integration/PokerProtocol/Observer/CommandObserver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace PokerProtocol.Observer
{
    public abstract class CommandObserver
    {
        public event EventHandler<StringEventArgs> CommandReceived = delegate { };

        protected abstract void receiveSomething(string line);

        public virtual void messageReceived(string line)
        {
            if (line == null)
            {
                return;
            }
            CommandReceived(this, new StringEventArgs(line));
            receiveSomething(line);
        }
    }
}
